File: Packages/StrandAnalytics/Sources/StrandAnalytics/SleepAwareStepCounter.swift
import Foundation
import WhoopProtocol

/// Sleep-context twin of Kotlin `SleepAwareStepCounter`. Counter decisions stay orientation-independent.
public enum SleepAwareStepCounter {
    public static let maxSecondsBetweenGaitDeltas = 3
    public static let minGaitBoutDurationSeconds = 4
    public static let minGaitBoutActiveSamples = 5
    public static let minGaitBoutTicks = 6

    public struct Count: Equatable, Sendable {
        public let totalTicks: Int
        public let acceptedOutsideSleepTicks: Int
        public let acceptedAwakeGapTicks: Int
        public let acceptedSleepBoutTicks: Int
        public let rejectedIsolatedSleepTicks: Int
        public let rejectedActivityClassTicks: Int
        public let rejectedImplausibleTicks: Int
        public let gravitySamplesAvailable: Int
        public let auxSamplesAvailable: Int

        public static let empty = Count(totalTicks: 0, acceptedOutsideSleepTicks: 0,
            acceptedAwakeGapTicks: 0, acceptedSleepBoutTicks: 0, rejectedIsolatedSleepTicks: 0,
            rejectedActivityClassTicks: 0, rejectedImplausibleTicks: 0,
            gravitySamplesAvailable: 0, auxSamplesAvailable: 0)

        /// Kotlin twin: `SleepAwareStepCounter.Count.plus`.
        public func adding(_ other: Count) -> Count {
            Count(totalTicks: totalTicks + other.totalTicks,
                acceptedOutsideSleepTicks: acceptedOutsideSleepTicks + other.acceptedOutsideSleepTicks,
                acceptedAwakeGapTicks: acceptedAwakeGapTicks + other.acceptedAwakeGapTicks,
                acceptedSleepBoutTicks: acceptedSleepBoutTicks + other.acceptedSleepBoutTicks,
                rejectedIsolatedSleepTicks: rejectedIsolatedSleepTicks + other.rejectedIsolatedSleepTicks,
                rejectedActivityClassTicks: rejectedActivityClassTicks + other.rejectedActivityClassTicks,
                rejectedImplausibleTicks: rejectedImplausibleTicks + other.rejectedImplausibleTicks,
                gravitySamplesAvailable: gravitySamplesAvailable + other.gravitySamplesAvailable,
                auxSamplesAvailable: auxSamplesAvailable + other.auxSamplesAvailable)
        }
    }

    /// Stateful, overlap-safe counter for database windows read in ascending pages.
    public final class Accumulator {
        private let sessions: [SleepSession]
        private let hasClasses: Bool
        private var previous: StepSample?
        private var outside = 0, awake = 0, sleep = 0, rejectedSleep = 0
        private var rejectedClass = 0, rejectedImplausible = 0
        private var gravityAvailable = 0, auxAvailable = 0
        private var pending: [(ts: Int, ticks: Int)] = []
        private var finished = false

        public init(sleepSessions: [SleepSession], hasActivityClasses: Bool) {
            sessions = sleepSessions.sorted { $0.start < $1.start }
            hasClasses = hasActivityClasses
        }

        /// Kotlin twin: `SleepAwareStepCounter.Accumulator.acceptPage`.
        @discardableResult public func acceptPage(_ samples: [StepSample]) -> Accumulator {
            precondition(!finished)
            for current in samples.sorted(by: { $0.ts < $1.ts }) {
                guard let prior = previous else { previous = current; continue }
                guard current.ts > prior.ts else { continue }
                previous = current
                let delta = (current.counter - prior.counter) & 0xffff
                guard StepsCounter.shouldCountDelta(activityClass: current.activityClass,
                                                     hasActivityClasses: hasClasses) else {
                    rejectedClass += delta; continue
                }
                guard StepsCounter.isPlausibleDelta(previousTs: prior.ts, currentTs: current.ts,
                                                    delta: delta) else {
                    rejectedImplausible += delta; continue
                }
                switch SleepAwareStepCounter.context(current.ts, sessions: sessions) {
                case 0: flush(); outside += delta
                case 1: flush(); awake += delta
                default:
                    if let last = pending.last, current.ts - last.ts > maxSecondsBetweenGaitDeltas { flush() }
                    pending.append((current.ts, delta))
                }
            }
            return self
        }

        /// Kotlin twin: `SleepAwareStepCounter.Accumulator.observeMotionPage`.
        @discardableResult public func observeMotion(gravityCount: Int, auxCount: Int) -> Accumulator {
            precondition(!finished)
            gravityAvailable += gravityCount
            auxAvailable += auxCount
            return self
        }

        /// Kotlin twin: `SleepAwareStepCounter.Accumulator.finish`.
        public func finish() -> Count {
            if !finished { flush(); finished = true }
            return Count(totalTicks: outside + awake + sleep, acceptedOutsideSleepTicks: outside,
                         acceptedAwakeGapTicks: awake, acceptedSleepBoutTicks: sleep,
                         rejectedIsolatedSleepTicks: rejectedSleep,
                         rejectedActivityClassTicks: rejectedClass,
                         rejectedImplausibleTicks: rejectedImplausible,
                         gravitySamplesAvailable: gravityAvailable,
                         auxSamplesAvailable: auxAvailable)
        }

        /// Kotlin twin: `SleepAwareStepCounter.Accumulator.flushSleepBout`.
        private func flush() {
            guard !pending.isEmpty else { return }
            let ticks = pending.reduce(0) { $0 + $1.ticks }
            let duration = pending.last!.ts - pending.first!.ts
            let coherent = pending.count >= minGaitBoutActiveSamples
                && duration >= minGaitBoutDurationSeconds && ticks >= minGaitBoutTicks
            if coherent { sleep += ticks } else { rejectedSleep += ticks }
            pending.removeAll(keepingCapacity: true)
        }
    }

    /// Kotlin twin: `SleepAwareStepCounter.stepsInWindow`.
    public static func stepsInWindow(_ samples: [StepSample], sleepSessions: [SleepSession]) -> Int? {
        let count = count(samples, sleepSessions: sleepSessions)
        return count.totalTicks > 0 ? count.totalTicks : nil
    }

    /// Kotlin twin: `SleepAwareStepCounter.count`.
    public static func count(_ samples: [StepSample], sleepSessions: [SleepSession]) -> Count {
        let sorted = samples.sorted { $0.ts < $1.ts }
        return Accumulator(sleepSessions: sleepSessions,
                           hasActivityClasses: StepsCounter.hasActivityClasses(sorted))
            .acceptPage(sorted).finish()
    }

    /// Kotlin twin: `SleepAwareStepCounter.contextAt`.
    private static func context(_ ts: Int, sessions: [SleepSession]) -> Int {
        guard let session = sessions.first(where: { ts >= $0.start && ts < $0.end }) else { return 0 }
        if let stage = session.stages.first(where: { ts >= $0.start && ts < $0.end }),
           SleepStageVocabulary.isWake(stage.stage) { return 1 }
        return 2
    }
}

File: android/app/src/main/java/com/noop/analytics/SleepAwareStepCounter.kt
package com.noop.analytics

import com.noop.data.GravitySample
import com.noop.data.StepSample
import com.noop.data.V18AuxRow

/**
 * Sleep-context wrapper around [StepsCounter]. It keeps the incumbent counter/class/rate rules, but does
 * not accept an isolated strap "walk" tick as proof that a sleeping person walked. Inside a scored sleep
 * stage, valid locomotion deltas must form a short, temporally coherent bout. An explicitly scored wake
 * gap is intentionally treated like ordinary awake time, so getting up for a child or the toilet counts.
 *
 * This kernel returns raw counter ticks exactly as [StepsCounter] does. It deliberately applies neither a
 * learned conversion factor nor a fixed wrist orientation: a strap may be worn on either arm, a limb or
 * clothing. Gravity/dynamic-acceleration and v18 auxiliary data are accepted for observability and future
 * shadow validation, but are not yet trusted as production gates. When either stream is absent, the
 * conservative temporal bout rule remains deterministic and fail-safe.
 */
object SleepAwareStepCounter {
    /** A one- or two-second bed movement cannot satisfy all three bout floors. */
    const val MAX_SECONDS_BETWEEN_GAIT_DELTAS: Long = 3L
    const val MIN_GAIT_BOUT_DURATION_SECONDS: Long = 4L
    const val MIN_GAIT_BOUT_ACTIVE_SAMPLES: Int = 5
    const val MIN_GAIT_BOUT_TICKS: Int = 6

    data class Count(
        val totalTicks: Int,
        val acceptedOutsideSleepTicks: Int,
        val acceptedAwakeGapTicks: Int,
        val acceptedSleepBoutTicks: Int,
        val rejectedIsolatedSleepTicks: Int,
        val rejectedActivityClassTicks: Int,
        val rejectedImplausibleTicks: Int,
        /** Availability only; neither raw stream changes the production decision yet. */
        val gravitySamplesAvailable: Int,
        val auxSamplesAvailable: Int,
    ) {
        operator fun plus(other: Count): Count = Count(
            totalTicks + other.totalTicks,
            acceptedOutsideSleepTicks + other.acceptedOutsideSleepTicks,
            acceptedAwakeGapTicks + other.acceptedAwakeGapTicks,
            acceptedSleepBoutTicks + other.acceptedSleepBoutTicks,
            rejectedIsolatedSleepTicks + other.rejectedIsolatedSleepTicks,
            rejectedActivityClassTicks + other.rejectedActivityClassTicks,
            rejectedImplausibleTicks + other.rejectedImplausibleTicks,
            gravitySamplesAvailable + other.gravitySamplesAvailable,
            auxSamplesAvailable + other.auxSamplesAvailable,
        )

        companion object {
            val EMPTY = Count(0, 0, 0, 0, 0, 0, 0, 0, 0)
        }
    }

    /** Preserve scored stages while edited leading/trailing edges remain conservatively inside sleep. */
    fun withBounds(session: DetectedSleep, start: Long, end: Long): DetectedSleep =
        session.copy(start = start, end = end)

    /** Effective bounds + stage content affecting exactly this cycle; unrelated sleep edits stay out. */
    fun contextSignature(sessions: List<DetectedSleep>, onset: Long, endExclusive: Long): String =
        sessions.asSequence()
            .filter { it.end > onset && it.start < endExclusive }
            .sortedWith(compareBy<DetectedSleep> { it.start }.thenBy { it.end })
            .joinToString("|") { session ->
                buildString {
                    append(session.start).append('-').append(session.end)
                    for (stage in session.stages.sortedBy { it.start }) {
                        append(':').append(stage.start).append('-').append(stage.end).append('=').append(stage.stage)
                    }
                }
            }

    /** Same nullable convention as [StepsCounter.stepsInWindow]: no accepted movement is `null`. */
    fun stepsInWindow(
        samples: List<StepSample>,
        sleepSessions: List<DetectedSleep>,
        gravity: List<GravitySample> = emptyList(),
        aux: List<V18AuxRow> = emptyList(),
    ): Int? = count(samples, sleepSessions, gravity, aux).totalTicks.takeIf { it > 0 }

    /** Detailed form intended for the Test Centre/shadow trace. */
    fun count(
        samples: List<StepSample>,
        sleepSessions: List<DetectedSleep>,
        gravity: List<GravitySample> = emptyList(),
        aux: List<V18AuxRow> = emptyList(),
    ): Count {
        val sorted = samples.sortedBy { it.ts }
        return Accumulator(sleepSessions, StepsCounter.hasActivityClasses(sorted))
            .observeMotionPage(gravity, aux)
            .acceptPage(sorted)
            .finish()
    }

    /**
     * Page-safe form for database windows larger than an in-memory query limit. The caller determines
     * [hasActivityClasses] once for the complete window (for example with an EXISTS query) and keeps one
     * accumulator for every ascending page. The previous counter sample and an unfinished in-sleep gait
     * bout survive page boundaries, so paging cannot change the answer.
     *
     * Pages may overlap at their boundary: timestamps already consumed are ignored. Calling [finish]
     * closes the last gait bout and makes the accumulator immutable.
     */
    class Accumulator(
        sleepSessions: List<DetectedSleep>,
        private val hasActivityClasses: Boolean,
    ) {
        private val sessions = sleepSessions.sortedBy { it.start }
        private var previous: StepSample? = null
        private var outside = 0
        private var awakeGap = 0
        private var sleepBout = 0
        private var rejectedSleep = 0
        private var rejectedClass = 0
        private var rejectedImplausible = 0
        private var gravityAvailable = 0
        private var auxAvailable = 0
        private var finished = false
        private val pendingSleepBout = ArrayList<Delta>()

        fun observeMotionPage(
            gravity: List<GravitySample> = emptyList(),
            aux: List<V18AuxRow> = emptyList(),
        ): Accumulator = apply {
            check(!finished) { "Accumulator is already finished" }
            gravityAvailable += gravity.size
            auxAvailable += aux.size
        }

        fun acceptPage(samples: List<StepSample>): Accumulator = apply {
            check(!finished) { "Accumulator is already finished" }
            for (current in samples.sortedBy { it.ts }) {
                val prior = previous
                if (prior != null && current.ts <= prior.ts) continue
                previous = current
                if (prior == null) continue

                val delta = (current.counter - prior.counter) and 0xFFFF
                if (!StepsCounter.shouldCountDelta(current.activityClass, hasActivityClasses)) {
                    rejectedClass += delta
                    continue
                }
                if (!StepsCounter.isPlausibleDelta(prior.ts, current.ts, delta)) {
                    rejectedImplausible += delta
                    continue
                }

                when (contextAt(current.ts, sessions)) {
                    Context.OUTSIDE_SLEEP -> {
                        flushSleepBout()
                        outside += delta
                    }
                    Context.AWAKE_GAP -> {
                        flushSleepBout()
                        awakeGap += delta
                    }
                    Context.SCORED_SLEEP -> {
                        val last = pendingSleepBout.lastOrNull()
                        if (last != null && current.ts - last.ts > MAX_SECONDS_BETWEEN_GAIT_DELTAS) {
                            flushSleepBout()
                        }
                        pendingSleepBout += Delta(current.ts, delta)
                    }
                }
            }
        }

        fun finish(): Count {
            if (!finished) {
                flushSleepBout()
                finished = true
            }
            return Count(
                totalTicks = outside + awakeGap + sleepBout,
                acceptedOutsideSleepTicks = outside,
                acceptedAwakeGapTicks = awakeGap,
                acceptedSleepBoutTicks = sleepBout,
                rejectedIsolatedSleepTicks = rejectedSleep,
                rejectedActivityClassTicks = rejectedClass,
                rejectedImplausibleTicks = rejectedImplausible,
                gravitySamplesAvailable = gravityAvailable,
                auxSamplesAvailable = auxAvailable,
            )
        }

        private fun flushSleepBout() {
            if (pendingSleepBout.isEmpty()) return
            val ticks = pendingSleepBout.sumOf { it.ticks }
            val duration = pendingSleepBout.last().ts - pendingSleepBout.first().ts
            val coherent = pendingSleepBout.size >= MIN_GAIT_BOUT_ACTIVE_SAMPLES &&
                duration >= MIN_GAIT_BOUT_DURATION_SECONDS &&
                ticks >= MIN_GAIT_BOUT_TICKS
            if (coherent) sleepBout += ticks else rejectedSleep += ticks
            pendingSleepBout.clear()
        }
    }

    private data class Delta(val ts: Long, val ticks: Int)
    private enum class Context { OUTSIDE_SLEEP, AWAKE_GAP, SCORED_SLEEP }

    private fun contextAt(ts: Long, sessions: List<DetectedSleep>): Context {
        val session = sessions.firstOrNull { ts >= it.start && ts < it.end } ?: return Context.OUTSIDE_SLEEP
        val stage = session.stages.firstOrNull { ts >= it.start && ts < it.end }
        return if (stage != null && SleepStageVocabulary.isWake(stage.stage)) {
            Context.AWAKE_GAP
        } else {
            // Missing/gapped staging is not evidence of wakefulness. This is intentionally conservative.
            Context.SCORED_SLEEP
        }
    }
}
